### PR TITLE
Transport queue

### PR DESCRIPTION
**Pull request by hellysmile**
Hey there!

* fixed deprecated `async` keyword https://www.python.org/dev/peps/pep-0492/#deprecation-plans
* bump to aiohttp 2.2+
* refactored `AioHttpTransport` to use queues for sending message. When queue is full it throws oldest message and logs error. Default behaviour is same as was before this patch. Passing `background_workers > 0` turns on queue mode.
* replaced `asyncio.wait_for` by `async_timeout.timeout`
* fixed deadly cache for dns
* added `close()` method for graceful transport shutdown
* `response.release()` is no longer coroutine and should be used in `finally` statement
* disabled built-n aiohttp timeout, cuz there is own top level should be used
* do not swallow `asyncio.CancelledError`
* close on queue mode waits until internal queue is empty, and in case of default mode - until all running messages are sent, as well there is timeout for close method

General idea is to prevent spawning tons of background tasks when application throwing a lot of messages

### CODE REVIEW

File: raven_aiohttp.py
Comment (asvetlov) [inline raven_aiohttp.py:21]: Why `asyncio.async` is not satisfactory?

File: raven_aiohttp.py
Comment (asvetlov) [inline raven_aiohttp.py]: Unused variable

File: raven_aiohttp.py
Comment (asvetlov) [inline raven_aiohttp.py]: Make optional params keyword-only?

Comment (asvetlov) [reply raven_aiohttp.py]: `return_exceptions` is bad name: `close()` doesn't return anything.
The behavior is much closer to *suppress*.
BTW in what situations suppressing is not desirable?

File: raven_aiohttp.py
Comment (asvetlov) [inline raven_aiohttp.py]: Cancel all not finished yet tasks/workers?

File: raven_aiohttp.py
Comment (asvetlov) [inline raven_aiohttp.py]: Doc needs recommendation about `background_workers` and `queue_maxsize` values.

File: raven_aiohttp.py
Comment (asvetlov) [inline raven_aiohttp.py]: Very non-obvious exception message.
Maybe we need dedicated exception type for this?

**Review by asvetlov — COMMENTED** [review]
Looks good but such complex functionality needs tests.

@mitsuhiko @dcramer could you give your feedback on proposed changes?

File: raven_aiohttp.py
Comment (hellysmile) [reply raven_aiohttp.py:21]: https://www.python.org/dev/peps/pep-0492/#deprecation-plans

File: raven_aiohttp.py
Comment (hellysmile) [reply raven_aiohttp.py]: right now it does same for any unexpected error
https://github.com/getsentry/raven-aiohttp/blob/master/raven_aiohttp.py#L99

how dedicated error can help in such case?

**Comment by mitsuhiko** [comment]
@ashwoods can you have a look at this?

File: raven_aiohttp.py
Comment (hellysmile) [reply raven_aiohttp.py]: Tasks not cancelled, we are waiting for them to be finished, a well same for workers.
If You want to stop it immediately `timeout=0` should be used

File: raven_aiohttp.py
Comment (hellysmile) [reply raven_aiohttp.py]: What about new custom exception?

File: raven_aiohttp.py
Comment (asvetlov) [reply raven_aiohttp.py]: Ok

**Comment by hellysmile** [comment]
There is an idea to have 2 transports. One is "fire and forget" another with "queue and workers". Something like explicit choice for users, not just optional parameters to select mode. Is it makes sense?

File: raven_aiohttp.py
Comment (hellysmile) [reply raven_aiohttp.py]: What about to make default timeout '0' ?

**Comment by ashwoods** [comment]
@hellysmile how would we manage failure with fire and forget? We don't want errors ending up in the abyss. And what we really need is tests. I'll update the boilerplate tomorrow with pytest-aiohttp and asynctest to have something in place.

**Comment by asvetlov** [comment]
@ashwoods the PR adds two modes for the plugin.
1. Keep current API but cancel all unsent tasks on `await transport.close()` explicitly.
2. Setup a pool of senders, push a new raven record into the queue and process it by senders.

Maybe we need splitting a client into two classes: 
* one for *fire and forget* with reporting not processed events on `close()` 
* other one for pushing events into a queue and processing them by workers.

P.S.
I'm not Sentry expert but aiohttp maintainer.
I'd like to help on any question about aiohttp and Sentry integration.

**Comment by ashwoods** [comment]
@asvetlov Ok, now I got it. +1 for two classes, more explicit for end users (+docs needed). 
I've been going through testing, would for sure appretiate help with that.

**Comment by hellysmile** [comment]
@ashwoods current  raven-aiohttp works exactly like fire and forget. Errors comes to logs via `failure_cb`. But if loop is full of tasks and application stopped users will see tons of asyncio errors like https://docs.python.org/3/library/asyncio-dev.html#asyncio-pending-task-destroyed 
Thats why explicit method close is added. Users can wait until all tasks done or just cancel them bypassing `timeout=0`

I'll make test suite, no worries.

**Comment by ashwoods** [comment]
@hellysmile many thx

File: raven_aiohttp.py
Comment (asvetlov) [reply raven_aiohttp.py:21]: ok

File: raven_aiohttp.py
Comment (asvetlov) [reply raven_aiohttp.py]: Not sure.
After stopping the server sending all pending events to sentry server makes sense.
It should be pretty fast procedure.

File: raven_aiohttp.py
Comment (asvetlov) [reply raven_aiohttp.py]: +1

**Comment by hellysmile** [comment]
@ashwoods @asvetlov basic test smoke tests + test boilerplate were added.

Lets take a look on tests structure and I'll start splitting code in 2 transports

File: raven_aiohttp.py
Comment (hellysmile) [inline raven_aiohttp.py]: @asvetlov lets place there `failure_cb(RuntimeError('AioHttpTransport is closing'))` ?

Comment (asvetlov) [reply raven_aiohttp.py]: Yes please

**Comment by asvetlov** [comment]
Looks good

**Comment by hellysmile** [comment]
Before adding ~100% test coverage would be great to hear feedback regarding provided functionality and code structure.

@ashwoods can You take a look please, thnx

**Comment by ashwoods** [comment]
Code looks awesome, thx! Needs class docstrings and docs for the abstract methods, but let's work on improving docs in a separate pull request. I'll do some testing this evening.

**Comment by hellysmile** [comment]
Thnx! Give me 1-2 days to finish tests coverage. Then we do a separate pull request for improving docs.

**Comment by hellysmile** [comment]
I've mostly done with tests, found interesting case, that `async_timeout.timeout` do not support `timeout=0`, which can be useful for this PR for instant closing of tasks\workers. Current implementation depends on https://github.com/aio-libs/async-timeout/issues/27 , or `asyncio.wait_for` can be used instead of `async_timeout.timeout`

**Comment by hellysmile** [comment]
`wait_for` with `timeout=0` did not work as expected for me. Here is minor fix to `asyncio` itself https://github.com/python/cpython/pull/3703 . Will wait untill final decision regarding `wait_for` . Then I'll try to adapt `async_timeout` to work same as `wait_for` only afterwards it can be used here

**Comment by hellysmile** [comment]
@asvetlov @ashwoods I think I am done with tests and docs, can You take a look on it?

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Good to me.
@ashwoods ?

**Comment by ashwoods** [comment]
Thx for the awesome work!

**Comment by Kentzo** [comment]
May I ask why `aiohttp.ClientSession` is re-created for every send unless `keepalive` is set to `True`?

**Comment by asvetlov** [comment]
I think it was in very early version.
IIRC I wrote this code.
`ClientSession` is useful for keep-aliving mode only, otherwise the price of new TCP connection establishment overweights all extra times for creating/deleting the session.

**Comment by Kentzo** [comment]
But there are other caches associated with ClientSession. I guess it can be further extended with future releases of aiohttp (etag?).

In this case, that bit looked confusing to me: why the hassle of an extra if/else if it makes sense for keepalive and irrelevant otherwise?

**Comment by Kentzo** [comment]
@hellysmile 

> fixed deadly cache for dns

Could you further elaborate about the issue and point to the the commit or relevant part of code that fixed it? As far as I can tell `aiohttp.TCPConnector` is still created with `use_dns_cache=True`.

**Comment by hellysmile** [comment]
Hey @Kentzo , `use_dns_cache` is no longer up to date change note. There was so many changes and a lot of fixes during development. Pull request description is out of date. Ill try to keep description up to date next time. Sorry..

**Comment by Kentzo** [comment]
@hellysmile That also made into CHANGES:

> - disabled built-in aiohttp dns cache

I think it should be fixed there then.

**Comment by hellysmile** [comment]
done, https://github.com/getsentry/raven-aiohttp/pull/23

**Pull request merged**
